### ISSUE TITLE
Incorrect results with Piecewise expression MatrixSymbol substitutions.

### ISSUE THREAD

**Issue by jmansour** [sympy/sympy#22175]
This follows on from #14933. Feel free to delete if it's preferred that that issue be reopened. 

I'm encountering strange results with the `master` branch version of Sympy:

```python
from sympy import Piecewise, Abs, Symbol, MatrixSymbol
x = Symbol("x")
fn = Piecewise((1.0, Abs(x) < 0.5), (3., (Abs(x - 2) <= 0.5) | (Abs(x- 1) < 0.5)), (1.0, True))
print(fn)
# Piecewise((1.0, Abs(x) < 0.5), (3.0, (x <= 2.5) & (x > 0.5)), (1.0, True))
mat_x = MatrixSymbol(  'mat_x', 1,1)
subfn = fn.xreplace({x:mat_x[0,0],})
print(subfn)
# 1.0
```
Against `sympy==1.8` this test produces the `TypeError: 'Dummy' object is not subscriptable` error discussed in #14933. 

Thanks

**Comment by oscarbenjamin** [comment]
> What we need in `as_set` (and perhaps elsewhere) is a way to identify the generators of the expression like `Poly` does:

The way that Poly does this is not quite right:
```python
In [1]: (exp(x) > 1).as_set()
Out[1]: (0, ∞)

In [2]: Poly(exp(x))
Out[2]: Poly((exp(x)), exp(x), domain='ZZ')
```

I think we just need to make it so that `as_set` is not being used in a case like this:
https://github.com/sympy/sympy/issues/14933#issuecomment-925300998
Of the three tasks 2 were done in #22156. What remains is to remove the use of `as_set` at least when the symbol is not `NumberKind`.

**Comment by smichr** [comment]
```python
>>> from sympy.core.exprtools import unigen
>>> unigen(Function('f')(x, y) + 1)
f(x, y)
>>> unigen(IndexedBase('f')[x] + 1)
f[x]
>>> unigen(cos(x) - 1)
x
>>> unigen(MatrixSymbol('M',1,1)[0,0] + 1)
M[0, 0]
```
with the following diff
```diff
diff --git a/sympy/core/exprtools.py b/sympy/core/exprtools.py
index 45923fc8d2..a6d713405c 100644
--- a/sympy/core/exprtools.py
+++ b/sympy/core/exprtools.py
@@ -26,6 +26,62 @@ def _isnumber(i):
     return isinstance(i, (SYMPY_INTS, float)) or i.is_Number
 
 
+def unigen(expr):
+    """return the largest univariate generator in the expression
+    that is not a sum or product
+
+    EXAMPLES
+    ========
+
+    >>> from sympy.core.exprtools import unigen
+    >>> from sympy.abc import x, y
+    >>> from sympy import Function
+    >>> unigen(x - 1)
+    x
+    >>> unigen(x + y) is None
+    True
+    >>> f = Function('f')
+    >>> unigen(f(x) + 1/(f(x) - 1) < 0)
+    x
+    """
+    from sympy.core.basic import preorder_traversal
+    from sympy.core.symbol import Symbol
+    from sympy.core.function import AppliedUndef
+    if not isinstance(expr, Basic):
+        return
+    free = expr.free_symbols
+    assert free
+    if len(free) == 1:
+        x = free.pop()
+        if isinstance(x, Symbol):
+            return x
+    else:
+        from sympy.utilities.iterables import sift
+        elem, compound = sift(free, lambda x: len(x.free_symbols) == 1, binary=True)
+        if len(compound) == 1 and compound[0].free_symbols == set(elem):
+            return compound[0]
+    gens = set()
+    pot = preorder_traversal(expr)
+    for i in pot:
+        if (i.is_Add or i.is_Mul or i.is_Pow or
+                i.is_Function and not isinstance(i, AppliedUndef)):
+            continue
+        free = i.free_symbols
+        if not free:
+            pot.skip()
+            continue
+        try:
+            gens.add(i.as_independent(free.pop())[1])
+        except AttributeError:
+            pass
+    gens = list(reversed(list(ordered(gens))))
+    D = Dummy()
+    for i, g in enumerate(gens):
+        d = expr.xreplace({g: D})
+        if not any(d.has(_) for _ in gens[i + 1:]):
+            return g
+
+
 def _monotonic_sign(self):
     """Return the value closest to 0 that ``self`` may have if all symbols
     are signed and the result is uniformly the same sign for all values of symbols.
diff --git a/sympy/core/relational.py b/sympy/core/relational.py
index 71fd73bb88..a9036182a2 100644
--- a/sympy/core/relational.py
+++ b/sympy/core/relational.py
@@ -401,9 +401,12 @@ def _eval_as_set(self):
         # self is univariate and periodicity(self, x) in (0, None)
         from sympy.solvers.inequalities import solve_univariate_inequality
         from sympy.sets.conditionset import ConditionSet
-        syms = self.free_symbols
-        assert len(syms) == 1
-        x = syms.pop()
+        from sympy.core.exprtools import unigen
+        from sympy.core.symbol import Dummy, Symbol
+        x = unigen(self)
+        assert x
+        if not isinstance(x, Symbol):
+            return self.xreplace({x: Dummy()})._eval_as_set()
         try:
             xset = solve_univariate_inequality(self, x, relational=False)
         except NotImplementedError:
diff --git a/sympy/functions/elementary/piecewise.py b/sympy/functions/elementary/piecewise.py
index c50e1f1a55..34ce55581c 100644
--- a/sympy/functions/elementary/piecewise.py
+++ b/sympy/functions/elementary/piecewise.py
@@ -157,6 +157,8 @@ def eval(cls, *_args):
         If there is a single arg with a True condition, its
         corresponding expression will be returned.
         """
+        from sympy.core.symbol import Symbol
+        from sympy.core.exprtools import unigen
         from sympy.functions.elementary.complexes import im, re
 
         if not _args:
@@ -171,20 +173,24 @@ def eval(cls, *_args):
         args = []
         for e, c in _args:
             if (not c.is_Atom and not isinstance(c, Relational) and
-                not c.has(im, re)):
-                free = c.free_symbols
-                if len(free) == 1:
-                    funcs = [i for i in c.atoms(Function)
-                             if not isinstance(i, Boolean)]
-                    if len(funcs) == 1 and len(
-                            c.xreplace({list(funcs)[0]: Dummy()}
-                            ).free_symbols) == 1:
-                        # we can treat function like a symbol
-                        free = funcs
+                    not c.has(im, re)):
+                x = unigen(c)
+                if x is None:
+                    d = None
+                elif not isinstance(x, Symbol):
+                    d = Dummy()
+                    x = unigen(c)
+                else:
+                    d = x
+                    g = x
+                if d:
                     _c = c
-                    x = free.pop()
                     try:
-                        c = c.as_set().as_relational(x)
+                        # `as_set` uses unigen but we can't
+                        # always recreate the relational with
+                        # a non-Symbol generator
+                        c = c.xreplace({g: d}).as_set(
+                            ).as_relational(d).xreplace({d: g})
                     except NotImplementedError:
                         pass
                     else:
```

**Comment by oscarbenjamin** [comment]
I still think that the ultimate fix here is just not using `as_set` rather than adding more machinery around the `as_set` call. With this diff
```diff
diff --git a/sympy/functions/elementary/piecewise.py b/sympy/functions/elementary/piecewise.py
index c50e1f1..b1b1d4d 100644
--- a/sympy/functions/elementary/piecewise.py
+++ b/sympy/functions/elementary/piecewise.py
@@ -1,5 +1,6 @@
 from sympy.core import Basic, S, Function, diff, Tuple, Dummy
 from sympy.core.basic import as_Basic
+from sympy.core.kind import NumberKind
 from sympy.core.numbers import Rational, NumberSymbol
 from sympy.core.relational import (Equality, Unequality, Relational,
     _canonical)
@@ -173,7 +174,7 @@ def eval(cls, *_args):
             if (not c.is_Atom and not isinstance(c, Relational) and
                 not c.has(im, re)):
                 free = c.free_symbols
-                if len(free) == 1:
+                if len(free) == 1 and all(x.kind is NumberKind for x in free):
                     funcs = [i for i in c.atoms(Function)
                              if not isinstance(i, Boolean)]
                     if len(funcs) == 1 and len(
```
We get:
```python
In [2]: fn
Out[2]: 
⎧1.0      for │x│ < 0.5    
⎪                          
⎨3.0  for x ≤ 2.5 ∧ x > 0.5
⎪                          
⎩1.0        otherwise      

In [3]: fn.xreplace({x:mat_x[0,0],})
Out[3]: 
⎧1.0         for │matₓ ₀₀│ < 0.5       
⎪                                      
⎨3.0  for matₓ ₀₀ ≤ 2.5 ∧ matₓ ₀₀ > 0.5
⎪                                      
⎩1.0              otherwise   
```
Restricting the use of `as_set` to `NumberKind` is an improvement but actually I just think that `as_set` shouldn't be used at all in `eval`. At best it could be used as part of simplification but only in a controlled way.

**Comment by oscarbenjamin** [comment]
When the Kind system was introduced the prime motivation was to be able to resolve problems like this. Lots of code assumes that expressions or symbols represent numbers but there didn't used to be a way to check that: https://github.com/sympy/sympy/issues/20158#issuecomment-735230658

**Comment by oscarbenjamin** [comment]
There have been a number of changes to Piecewise since the last comment here and this seems to be fixed now:
```python
In [1]: from sympy import Piecewise, Abs, Symbol, MatrixSymbol
   ...: 
   ...: x = Symbol("x")
   ...: fn = Piecewise(
   ...:     (1.0, Abs(x) < 0.5), (3.0, (Abs(x - 2) <= 0.5) | (Abs(x - 1) < 0.5)), (1.0, True)
   ...: )
   ...: print(fn)
   ...: # Piecewise((1.0, Abs(x) < 0.5), (3.0, (x <= 2.5) & (x > 0.5)), (1.0, True))
   ...: mat_x = MatrixSymbol("mat_x", 1, 1)
   ...: subfn = fn.xreplace(
   ...:     {
   ...:         x: mat_x[0, 0],
   ...:     }
   ...: )
   ...: print(subfn)
Piecewise((1.0, Abs(x) < 0.5), (3.0, (Abs(x - 2) <= 0.5) | (Abs(x - 1) < 0.5)), (1.0, True))
Piecewise((1.0, Abs(mat_x[0, 0]) < 0.5), (3.0, (Abs(mat_x[0, 0] - 2) <= 0.5) | (Abs(mat_x[0, 0] - 1) < 0.5)), (1.0, True))
```

**Comment by oscarbenjamin** [comment]
Seems this was fixed in df4e2f8eb12fa78002f0b78d15b6e5294dbf7b09 from #22156 by @oscargus which also explicitly mentions #14933.

I'll close this. Please say if anything about this does not seem to be fixed (the fix is in 1.10 which is already released).
